swim/heal_via_discover_provider.go

```
@@ -0,0 +1,168 @@
+	return &discoverProviderHealer{
+		node:             n,
+		baseProbabillity: baseProbability,
+		period:           period,
+		logger:           logging.Logger("healer").WithField("local", n.Address()),
+	}
+}
+
+// Start the partition healing loop
+func (h *discoverProviderHealer) Start() {
+	h.quit = make(chan struct{})
+	go func() {
+		for {
+			// attempt heal with the pro
+			if rand.Float64() < h.Probability() {
+				h.Heal()
+			}
+
+			// loop or quit
+			select {
+			case <-h.node.clock.After(h.period):
+			case <-h.quit:
+				return
+			}
+		}
+	}()
+}
+
+// Stop the partition healing loop.
+func (h *discoverProviderHealer) Stop() {
+	if h.quit == nil {
+		return
+	}
+
+	select {
+	case <-h.quit: // channel already closed, do nothing
+	default:
+		close(h.quit)
+	}
+}
+
+// Probability returns the probability when a heal should be attempted
+// we want to throttle the heal attempts to alleviate pressure on the
+// discover provider.
+func (h *discoverProviderHealer) Probability() float64 {
+	// avoid division by zero.
+	if h.previousHostListSize < h.node.CountReachableMembers() {
+		h.previousHostListSize = h.node.CountReachableMembers()
+	}
+	if h.previousHostListSize < 1 {
+		h.previousHostListSize = 1
+	}
+	return h.baseProbabillity / float64(h.previousHostListSize)
+}
+
+// Heal iterates over the hostList that the discoverProvider provides. If the
+// node encounters a host that is faulty or not in the membership, we pick that
+// node as a target to perform a partition heal with.
+//
+// If heal was attempted, returns identities of the target nodes.
+func (h *discoverProviderHealer) Heal() []string {
+	h.node.emit(DiscoHealEvent{})
+	// get list from discovery provider
+	if h.node.discoverProvider == nil {
+		return []string{}
+	}
+	hostList, err := h.node.discoverProvider.Hosts()
+	if err != nil {
+		h.logger.Warn("healer unable to receive host list from discover provider")
```

The admin endpoint exposes a list of hosts to be returned, and an error. Can you propagate the error from here, so the query from the admin endpoint can see it too?

Reason: operator triggers a /heal, which returns a success and empty list of hosts, but actually fails.